packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt

```diff
@@ -239,15 +199,15 @@ internal object RealmObjectHelper {
         nullable: Boolean
     ): R? {
         obj.checkValid()
-        val propertyInfo = checkPropertyType(obj, propertyName, CollectionType.RLM_COLLECTION_TYPE_NONE, clazz, nullable)
+        val propertyInfo =
+            checkPropertyType(obj, propertyName, CollectionType.RLM_COLLECTION_TYPE_NONE, clazz, nullable)
+        val realmValue = getValueByKey(obj, propertyInfo.key)
+        // Consider moving this dynamic conversion to Converters.kt
         val value = when (clazz) {
-            RealmInstant::class -> getTimestamp<R>(obj, propertyName)
-            DynamicRealmObject::class -> getObjectByKey<DynamicRealmObject>(obj, propertyInfo.key)
-            DynamicMutableRealmObject::class -> getObjectByKey<DynamicMutableRealmObject>(
-                obj,
-                propertyInfo.key
-            )
-            else -> getValueByKey<R>(obj, propertyInfo.key)
+            DynamicRealmObject::class,
+            DynamicMutableRealmObject::class ->
+                realmValueToRealmObject(realmValue, clazz as KClass<out RealmObject>, obj.mediator, obj.owner)
+            else -> primitiveTypeConverters.getValue(clazz).realmValueToPublic(realmValue)
```

Comment:
Do we have a test for what happens if no converter is found here?

Reply:
`getValue()` returns the `StaticPassThroughConverter` as supplied with `withDefault` when initializing `primitiveTypeConverters`. This means that tests for all the types without explicit converts in `primitiveTypeConverts` are already exercising this.